### ISSUE TITLE
Category is no longer optional

### ISSUE THREAD

**Issue by bhepple** [dvehrs/podget#45]
Please take it as a tribute to podget that I've had no need to upgrade for such a looong time - but in an idle moment I went from 0.7.6 to 0.8.5 and found that my `servers` file no longer works. 

After looking at the new code it seems that it breaks when there is no `category` or `feed names` - I had just the RSS URL.

I had to add '.' as a category to all my feeds which is OK, but it took me a while to figure out what was breaking - and I'm pretty good with bash. 

To save other people this hassle, maybe this fix would help to promote an empty category to '.':

```
*** podget	2019-05-03 08:58:08.581998449 +1000
--- podget.mod	2019-05-03 08:57:31.839797383 +1000
***************
*** 1875,1879 ****
  
              # Enable user configured no category.
!             if (( $(${EXPR} "${FEED_CATEGORY}" : "No_Category") > 0 )); then
                  # Set to single period which is a bashism for current directory.
                  FEED_CATEGORY="."
--- 1875,1879 ----
  
              # Enable user configured no category.
!             if [[ "${FEED_CATEGORY}" == "" ]] || (( $(${EXPR} "${FEED_CATEGORY}" : "No_Category") > 0 )); then
                  # Set to single period which is a bashism for current directory.
                  FEED_CATEGORY="."

```

**Comment by dvehrs** [comment]
This is an interesting idea and should work for cases where the feed category and name are both blank.  

The notes included with a new serverlist include directions to use either a single period or 'No_Category' for feeds that you wish to disable a category for.  However it seems you're not to different from myself and if I assume most users are too, then the notes portion of a new serverlist file will be quickly deleted soon after the file has been created.  So what we have here is an edge case that I had not anticipated but probably should have.  

To make matters worse, it appears that the manual page for Podget has not had it's EXAMPLE SERVER LIST section updated in a while.  A long while.  For example it does not include the advice to use either "No_Category" or a single period for feeds that you wish to disable the category for.  So that needs to be updated if only to help alleviate some of the frustration felt when an update to podget causes a serverlist to stop working.

Podget verifies that a users podgetrc file was made with a sufficiently new version so that all required options are specified.  Which raises the question, is it better to notify the user that they need to upgrade their serverlist (and how to do it) or is it better to just quietly correct this minor issue and proceed processing the serverlist?    I lean toward the latter option as I prefer to have fewer interruptions in peoples use of podget and I can understand why a user might wish to use podget in this fashion.    

From a testing perspective, this adds a few additional things to test for.  For example, what happens if you have blank feed category & name but also use one of the additional options (like OPT_CONTENT_DISPOSITION or OPT_FEED_ORDER_ASCENDING)? Or what happens if you have an atom feed with both blank?  And let me be clear, just because it makes testing a bit more complicated is never a good reason to not include a desirable feature.  I include this just to preemptively explain why it may take a little while to integrate this feature.  

The good news is I think your proposed fix will work, the no-so-good news is that there appear to be a few things that come along with this simple fix that may take a bit longer to address.  Fortunately your timing is good and I have some time I can dedicate to this in the near future.   Once i start I will branch a new version of the DEV branch to work on this and you can monitor my progress there.  

Thanks for the report and thank you for the kind words.  It has always been a goal of mine for podget to simply be something that did what the user wanted reliably.  I've always assumed that was true for a few users out there but never had any evidence that it was so.  It's also good to hear that an upgrade of a nearly 4 year old version mostly worked or at least it worked well enough that it was fixable.  Hopefully when we get your proposed fix integrated it will work even better for more users.

**Comment by dvehrs** [comment]
The branch for these changes is dev-optionalcategory.

So far I've updated the manual page to roughly reflect the current status of podget.  It was 2 yrs out of date, fortunately there wasn't too much to add.  

I think we're now ready to start trying your proposed fix and testing it.

**Comment by dvehrs** [comment]
OK, I've uploaded a test version.  

I did not use [[ "${FEED_CATEGORY}" == "" ]] because when nounset is disabled that will result in a positive answers for both empty or unset strings.  Now we have nounset enabled, so it will only trigger for empty strings.  However this use can be seen as vague and what we want here is a test we can use that will only trigger for empty strings regards of whether nounset is enabled or not.  That test is [[ -z "${FEED_CATEGORY-empty}" ]], where the - designates a default value to be set for the value for any unset variable so in this case the only zero length string we can have is if FEED_CATEGORY was set to the empty string "".  

I found three other places in podget where I was using the same comparison as you suggested and replaced those as well.

I've only tested with a couple feeds but the script mostly appears to work.  There are a couple small issues in how playlists are being created that need to be addressed.  When FEED_CATEGORY and FEED_NAME are blank, it lists the items on the playlist as ".//filename".  The first problem is the blank FEED_NAME between the two slashes.  That should probably be replaced with a period as we do for a blank FEED_CATEGORY.  Then once that is done, we should probably find a way to remove the unnecessary './' from each listing as many times as necessary.

**Comment by bhepple** [comment]
Looks good so far. I'll keep it running over the next few days and let you now if there are problems.

Cheers

Bob

**Comment by dvehrs** [comment]
I just pushed some new updates to dev-optionalcategory that allow for the use of the OPT_ items on a feed.  From basic testing it appears to work, but needs a bit more before I merge it into the dev branch.

**Comment by dvehrs** [comment]
I just finished testing with a wide variety of podcasts and it seems to work.  Unless you have any issues, I think we're ready to integrate these changes into the 'dev' branch.

**Comment by bhepple** [comment]
No issues at all - downloads every 5am like a charm.

Thanks


Bob

**Comment by dvehrs** [comment]
OK, I've merged the changes into the 'dev' branch.  A little more testing and we should be ready to release a new version.

**Comment by bhepple** [comment]
Thanks Dave,

I've installed the dev version and all seems well on my (fairly limited)
.podget

Cheers


Bob

**Comment by dvehrs** [comment]
Included in v0.8.6 release.  
Enjoy!
